jms/src/test/scala/akka/stream/alpakka/jms/scaladsl/JmsAckConnectorsSpec.scala

```diff
@@ -0,0 +1,371 @@
+
+      Thread.sleep(500)
+      println("Elements in resultQueue before first shutdown: " + resultQueue.size)
+
+      val killSwitch2 = jmsSource
+        .take(msgsIn.size - resultQueue.size)
+        .map { env =>
+          resultQueue.add(env.message.asInstanceOf[TextMessage].getText)
+          env.acknowledge()
+        }
+        .to(Sink.ignore)
+        .run()
+
+      val resultList = new mutable.ArrayBuffer[String](numsIn.size)
+
+      @tailrec
+      def keepPolling(): Unit =
+        Option(resultQueue.poll(2, TimeUnit.SECONDS)) match {
+          case Some(entry) =>
+            resultList += entry
+            keepPolling()
+          case None =>
+        }
+
+      keepPolling()
+
+      println("Elements in resultList now: " + resultList.size)
+      killSwitch2.shutdown()
+
+      resultList should contain theSameElementsAs numsIn.map(_.toString)
+    }
+
+    "ensure no message loss when aborting a stream" in withServer() { ctx =>
+      val url: String = ctx.url
+      val connectionFactory = new ActiveMQConnectionFactory(url)
+
+      val jmsSink: Sink[JmsTextMessage, NotUsed] = JmsSink(
+        JmsSinkSettings(connectionFactory).withQueue("numbers")
+      )
+
+      val numsIn = 1 to 100
+      val msgsIn = numsIn map { n =>
+        JmsTextMessage(n.toString).add("Number", n)
+      }
+
+      Source(msgsIn).runWith(jmsSink)
+
+      val jmsSource: Source[AckEnvelope, KillSwitch] = JmsSource.ackSource(
+        JmsSourceSettings(connectionFactory).withSessionCount(5).withBufferSize(0).withQueue("numbers")
+      )
+
+      val resultQueue = new LinkedBlockingQueue[String]()
+
+      val killSwitch = jmsSource
+        .take(msgsIn.size)
+        .throttle(10, 1.second, 2, ThrottleMode.shaping)
+        .map { env =>
+          resultQueue.add(env.message.asInstanceOf[TextMessage].getText)
+          env.acknowledge()
+        }
+        .to(Sink.ignore)
+        .run()
+
+      Thread.sleep(2000)
+      killSwitch.abort(new Exception("Test exception"))
+
+      Thread.sleep(500)
+      println("Elements in resultQueue before aborting: " + resultQueue.size)
+
+      val killSwitch2 = jmsSource
+        .take(msgsIn.size - resultQueue.size)
+        .map { env =>
+          resultQueue.add(env.message.asInstanceOf[TextMessage].getText)
+          env.acknowledge()
+        }
+        .to(Sink.ignore)
+        .run()
+
+      val resultList = new mutable.ArrayBuffer[String](numsIn.size)
+
+      @tailrec
+      def keepPolling(): Unit =
+        Option(resultQueue.poll(2, TimeUnit.SECONDS)) match {
+          case Some(entry) =>
+            resultList += entry
+            keepPolling()
+          case None =>
+        }
+
+      keepPolling()
+
+      println("Elements in resultList now: " + resultList.size)
+      killSwitch2.shutdown()
+
+      implicit val _ = Ordering.by { s: String =>
+        s.toInt
+      }
+      resultList.to[SortedSet] should contain theSameElementsAs numsIn.map(_.toString)
```

Comment:
No need to do ordering and use sorted set, `toSet should contain theSameElementsAs ...` is enough.

Reply:
That is correct, until you have a mismatch. And since re-delivery is always out of order, it is extremely hard to figure out what is missing. This is why I prefer to keep it sorted, just for the sanity of looking at the failed results when it does fail.